sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/ToNumberParser.scala

```diff
@@ -0,0 +1,517 @@
+    }
+    // Make sure that any dollar sign in the format string occurs before any digits.
+    else if (firstDigitIndex < firstDollarSignIndex) {
+      s"Currency characters must appear before digits in the number format: '$originNumberFormat'"
+    }
+    // Make sure that any dollar sign in the format string occurs before any decimal point.
+    else if (firstDecimalPointIndex != -1 && firstDecimalPointIndex < firstDollarSignIndex) {
+      "Currency characters must appear before any decimal point in the " +
+        s"number format: '$originNumberFormat'"
+    }
+    // Make sure that any thousands separators in the format string have digits before and after.
+    else if (digitGroupsBeforeDecimalPoint.exists {
+      case DigitGroups(tokens) =>
+        tokens.zipWithIndex.exists({
+          case (_: ThousandsSeparator, j: Int) if j == 0 || j == tokens.length - 1 =>
+            true
+          case (_: ThousandsSeparator, j: Int) if tokens(j - 1).isInstanceOf[ThousandsSeparator] =>
+            true
+          case (_: ThousandsSeparator, j: Int) if tokens(j + 1).isInstanceOf[ThousandsSeparator] =>
+            true
+          case _ =>
+            false
+        })
+    }) {
+      "Thousands separators (,) must have digits in between them " +
+        s"in the number format: '$originNumberFormat'"
+    }
+    // Thousands separators are not allowed after the decimal point, if any.
+    else if (digitGroupsAfterDecimalPoint.exists {
+      case DigitGroups(tokens) => tokens.exists(_.isInstanceOf[ThousandsSeparator])
+    }) {
+      "Thousands separators (,) may not appear after the decimal point " +
+        s"in the number format: '$originNumberFormat'"
+    }
+    // Validation of the format string finished successfully.
+    else {
+      ""
+    }
+  }
+
+  /**
+   * Convert string to numeric based on the given number format.
+   *
+   * Iterates through the [[inputTokens]] obtained from processing the format string, while also
+   * keeping a parallel index into the input string. Throws an exception if the latter does not
+   * contain expected characters at any point.
+   *
+   * @param input the string that needs to converted
+   * @return the result Decimal value obtained from string parsing
+   */
+  def parse(input: UTF8String): Decimal = {
+    val inputStr = input.toString.trim
+    val beforeDecimalPoint = new StringBuilder()
+    val afterDecimalPoint = new StringBuilder()
+    var reachedDecimalPoint = false
+    // Number of times that the input specified a negative result, such as with a minus sign.
+    var negateResult = 0
+    // This is an index into the characters of the provided input string.
+    var inputIndex = 0
+    // This is an index into the tokens of the provided format string.
+    var formatIndex = 0
+
+    // Iterate through the tokens representing the provided format string, in order.
+    while (formatIndex < inputTokens.size) {
```

Comment:
`inputTokens` looks a bit weird here, how about renaming it to `formatTokens`?

Reply:
Good idea, done.